src/main/java/com/google/jenkins/plugins/k8sengine/Manifests.java

```diff
@@ -63,16 +76,70 @@ public String getKind() {
     }
 
     /** @return The name. */
+    public Optional<String> getName() {
+      return getMetadata().isPresent()
+          ? Optional.of((String) getMetadata().get().get("name"))
+          : Optional.empty();
+    }
+
+    /**
+     * Ensures this {@link ManifestObject} has labels, modifying in-place as needed, finally
+     * returning the labels.
+     *
+     * @return The labels for this {@link ManifestObject}.
+     */
     @SuppressWarnings("unchecked")
-    public String getName() {
-      Map<String, Object> metadata = (Map<String, Object>) source.get("metadata");
-      return (String) metadata.get("name");
+    public Map<String, String> getOrCreateLabels() {
+      Map<String, Object> metadata = getOrCreateMetadata();
+
+      if (!metadata.containsKey("labels")) {
+        metadata.put("labels", new LinkedHashMap<String, String>());
+      }
+
+      return (Map<String, String>) metadata.get("labels");
+    }
+
+    /**
+     * Adds the specified label key and value to this {@link ManifestObject}'s metadata labels. Will
+     * ensure a label map exists upon execution.
+     *
+     * @param key The key of the label to be added.
+     * @param value The value of the label to be added.
+     */
+    public void addLabel(String key, String value) {
+      Map<String, String> labels = getOrCreateLabels();
+      Set<String> labelValues =
+          labels.get(key) != null
+              ? new HashSet<>(Arrays.asList(labels.get(key).split(",")))
```

Comment:
Wrapping this list (which wraps an array) in a set doesn't really seem necessary here since String.join accepts any Iterable. Also somewhat concerned about the order changing when iterating over a set vs. the original list.

Reply:
The set is used to ensure duplicates aren't added. Ordering is ignored in the spec.

Reply:
Thanks for clearing that up. Since it's not immediately clear could you add an inline comment explaining the need to remove duplicates?

Reply:
Sure thing :)